### ISSUE TITLE
Add SPM support by using extensible build tools

### ISSUE THREAD

**Issue by fjtrujy** [SwiftGen/SwiftGen#913]
Officially SPM will support extensible build tools in Swift 5.6. 
It is actually already enabled in Swift 5.5 by using `SPM_ENABLE_PLUGINS=1`

Here you have the official info, where they actually mention `SwiftGen` as one of the tools that can use this feature.
https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md

I think that we can start preparing our `Package plugins`, right?

Thanks

**Comment by djbe** [comment]
Nice! I couldn't find much about that env. flag, except that it may even not be needed anymore? https://twitter.com/bok_/status/1409480299382145028

I'd noticed there was a follow-up proposal (https://github.com/apple/swift-evolution/blob/main/proposals/0325-swiftpm-additional-plugin-apis.md), after which we could get started on this. From what I could gather, it changes the implementation a bit (and even provides a SwiftGen example 😅. There's also https://github.com/apple/swift-evolution/blob/main/proposals/0332-swiftpm-command-plugins.md, which we could use to provide commands that the user could run manually, such as lint config.

**Comment by fjtrujy** [comment]
> Nice! I couldn't find much about that env. flag, except that it may even not be needed anymore? https://twitter.com/bok_/status/1409480299382145028

I was thinking the same, but if you try to create a `Package` using the `plugin` you suffer a compilation error advising you to use the flag.

Actually here you have the compilation error for a POC that I did for `SwiftLint`

```bash
➜  SwiftLint git:(plugin) swift build
'SwiftLint' /Users/fjtrujy/Projects/SwiftLint: error: plugin target 'SwiftLintPlugin' cannot be used because the feature isn't enabled (set SWIFTPM_ENABLE_PLUGINS=1 in environment)
error: ExitCode(rawValue: 1)
[0/1] Planning buildwarning: failed to load the cached build description: fatalError
'SwiftLint' /Users/fjtrujy/Projects/SwiftLint: error: plugin target 'SwiftLintPlugin' cannot be used because the feature isn't enabled (set SWIFTPM_ENABLE_PLUGINS=1 in environment)
```

> I'd noticed there was a follow-up proposal (https://github.com/apple/swift-evolution/blob/main/proposals/0325-swiftpm-additional-plugin-apis.md), after which we could get started on this. From what I could gather, it changes the implementation a bit (and even provides a SwiftGen example 😅. There's also https://github.com/apple/swift-evolution/blob/main/proposals/0332-swiftpm-command-plugins.md, which we could use to provide commands that the user could run manually, such as lint config.

Yeah! I thought that probably you were aware of this feature already, but I couldn't find anything either in the issues or PRs.

**Comment by RafaelFernandezAlv** [comment]
The plugin feature is enabled in Xcode 13.3 beta 2. You only need change the `Package.swift` with the correct verison `// swift-tools-version: 5.6` and you can use it

**Comment by Kondamon** [comment]
I have tried to implement it according to [SE-0303](https://github.com/apple/swift-evolution/blob/main/proposals/0303-swiftpm-extensible-build-tools.md#example-1-swiftgen) and [SE-0305](https://github.com/apple/swift-evolution/blob/main/proposals/0305-swiftpm-binary-target-improvements.md)

However, I receive the error `Plugin does not have access to a tool named ‘swiftgen’'.` when I use the plugin in another swift package. The updated SwiftGen plugin doesn't show any errors in XCode. Perhaps there are any access control issues or still an issue with configuration.
https://github.com/SwiftGen/SwiftGen/compare/stable...Kondamon:stable

**Comment by drspoton** [comment]
I've implemented the plugin (https://github.com/drspoton/swiftgen-plugin-spm).
Would be great to have an option to clear the directory with generated files.
In case of changing output file name, it generates files in plugin work directory and it's pretty inconvenient to remove them manually

**Comment by nicorichard** [comment]
[I implemented a plug-in](https://github.com/nicorichard/SwiftGenPlugin) as well to accompany #926.

If my PR could be merged then @drspoton, @Kondamon, myself, and anyone else who wants to write a Swift Plug-in for SwiftGen could use the trusted binary generated from this repository.

**Comment by rzulkoski** [comment]
@nicorichard I was doing some testing with your plugin and was hitting some snags when attempting to build an iOS project that relies on a package which uses the plugin.

Here's my [post](https://forums.swift.org/t/plugin-command-not-running-when-package-is-indirectly-included-in-xcode-project/57168) on the Swift forums outlining my setup.

And here's my [sample project](https://github.com/rzulkoski/PluginNotRunning) demonstrating the issue.

Did you encounter anything like this in your testing?

**Comment by djbe** [comment]
I've created an official SwiftGenPlugin based on the work posted in this thread, and expanded upon it. You can find it here:
https://github.com/SwiftGen/SwiftGenPlugin

It provides:
- a pre-build plugin, to generate files on each build
- a generate command, to manually invoke SwiftGen

Note: currently it points to the artifact file from a prerelease tag. This will be changed to point to the **official** fork when it is released.

**Comment by djbe** [comment]
Implemented by SwiftGen 6.6.0, which has been released.
